packages/x-data-grid/src/components/base/GridOverlays.tsx

```diff
@@ -85,46 +92,29 @@ function GridOverlayWrapper(props: React.PropsWithChildren<{ overlayType: string
   );
 }
 
-GridOverlayWrapper.propTypes = {
+GridOverlays.propTypes = {
   // ----------------------------- Warning --------------------------------
   // | These PropTypes are generated from the TypeScript type definitions |
   // | To update them edit the TypeScript types and run "pnpm proptypes"  |
   // ----------------------------------------------------------------------
-  overlayType: PropTypes.string.isRequired,
+  loadingOverlayVariant: PropTypes.oneOf(['circular-progress', 'linear-progress', 'skeleton']),
+  overlayType: PropTypes.oneOf(['loadingOverlay', 'noResultsOverlay', 'noRowsOverlay']),
 } as any;
 
-export function GridOverlays() {
-  const apiRef = useGridApiContext();
+export function GridOverlays(props: GridOverlaysProps) {
+  const { overlayType } = props;
   const rootProps = useGridRootProps();
 
-  const totalRowCount = useGridSelector(apiRef, gridRowCountSelector);
-  const visibleRowCount = useGridSelector(apiRef, gridExpandedRowCountSelector);
-  const loading = useGridSelector(apiRef, gridRowsLoadingSelector);
-
-  const showNoRowsOverlay = !loading && totalRowCount === 0;
-  const showNoResultsOverlay = !loading && totalRowCount > 0 && visibleRowCount === 0;
-
-  let overlay: React.JSX.Element | null = null;
-  let overlayType = '';
-
-  if (showNoRowsOverlay) {
-    overlay = <rootProps.slots.noRowsOverlay {...rootProps.slotProps?.noRowsOverlay} />;
-    overlayType = 'noRowsOverlay';
-  }
-
-  if (showNoResultsOverlay) {
-    overlay = <rootProps.slots.noResultsOverlay {...rootProps.slotProps?.noResultsOverlay} />;
-    overlayType = 'noResultsOverlay';
-  }
-
-  if (loading) {
-    overlay = <rootProps.slots.loadingOverlay {...rootProps.slotProps?.loadingOverlay} />;
-    overlayType = 'loadingOverlay';
-  }
-
-  if (overlay === null) {
+  if (!overlayType) {
     return null;
   }
 
-  return <GridOverlayWrapper overlayType={overlayType}>{overlay}</GridOverlayWrapper>;
+  const Overlay = rootProps.slots?.[overlayType];
```

Comment:
@MBilalShafi yeah, it could do. Do you know why we export `GridOverlays` from the packages? If we are going to return the component and component props from the hook, the `GridOverlays` component would pretty redundant.

Reply:
> Do you know why we export GridOverlays from the packages? If we are going to return the component and component props from the hook, the GridOverlays component would pretty redundant.

I think it's a legacy export that could be gotten rid of in the next major if it does not make much sense anymore!

Reply:
> useGridOverlays could go one step further and directly return Overlay and overlayProps
Yeah, feels like we could refactor this part and remove the GridOverlays component for the next major

Reply:
Hey 👋,
I've opened https://github.com/mui/mui-x/pull/15573 to address this
Feel free to share your thoughts